crates/block-producer/src/mempool/inflight_state/mod.rs

```diff
@@ -0,0 +1,570 @@
+    accounts: BTreeMap<AccountId, InflightAccountState>,
+
+    /// Nullifiers produced by the input notes of inflight transactions.
+    nullifiers: BTreeSet<Nullifier>,
+
+    /// Notes created by inflight transactions.
+    ///
+    /// Some of these may already be consumed - check the nullifiers.
+    output_notes: BTreeMap<NoteId, OutputNoteState>,
+}
+
+/// The aggregated impact of a set of sequential transactions on the [InflightState].
+pub struct StateDelta {
+    /// The number of transactions that affected each account.
+    account_transactions: BTreeMap<AccountId, usize>,
+
+    /// The nullifiers consumed by the transactions.
+    nullifiers: BTreeSet<Nullifier>,
+
+    /// The notes produced by the transactions.
+    output_notes: BTreeSet<NoteId>,
+}
+
+impl StateDelta {
+    pub fn new(txs: &[AuthenticatedTransaction]) -> Self {
+        let mut account_transactions = BTreeMap::<AccountId, usize>::new();
+        let mut nullifiers = BTreeSet::new();
+        let mut output_notes = BTreeSet::new();
+
+        for tx in txs {
+            *account_transactions.entry(tx.account_id()).or_default() += 1;
+            nullifiers.extend(tx.nullifiers());
+            output_notes.extend(tx.output_notes());
+        }
+
+        Self {
+            account_transactions,
+            nullifiers,
+            output_notes,
+        }
+    }
+}
+
+impl InflightState {
+    /// Appends the transaction to the inflight state.
+    ///
+    /// This operation is atomic i.e. a rejected transaction has no impact of the state.
+    pub fn add_transaction(
+        &mut self,
+        tx: &AuthenticatedTransaction,
+    ) -> Result<BTreeSet<TransactionId>, VerifyTxError> {
+        // Separate verification and state mutation so that a rejected transaction
+        // does not impact the state (atomicity).
+        self.verify_transaction(tx)?;
+
+        let parents = self.insert_transaction(tx);
+
+        Ok(parents)
+    }
+
+    fn verify_transaction(&self, tx: &AuthenticatedTransaction) -> Result<(), VerifyTxError> {
+        // Ensure current account state is correct.
+        let current = self
+            .accounts
+            .get(&tx.account_id())
+            .and_then(|account_state| account_state.current_state())
+            .copied()
+            .or(tx.store_account_state());
+        let expected = tx.account_update().init_state_hash();
+
+        if expected != current.unwrap_or_default() {
+            return Err(VerifyTxError::IncorrectAccountInitialHash {
+                tx_initial_account_hash: expected,
+                current_account_hash: current,
+            });
+        }
+
+        // Ensure nullifiers aren't already present.
+        let double_spend = tx
+            .nullifiers()
+            .filter(|nullifier| self.nullifiers.contains(nullifier))
+            .collect::<Vec<_>>();
+        if !double_spend.is_empty() {
+            return Err(VerifyTxError::InputNotesAlreadyConsumed(double_spend));
+        }
```

Comment:
I would add a comment here explaining that we've already checked nullifiers against the store.